File: packages/vapi/android/src/main/java/io/capawesome/capacitorjs/plugins/vapi/VapiPlugin.java
package io.capawesome.capacitorjs.plugins.vapi;

import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import io.capawesome.capacitorjs.plugins.vapi.classes.options.SetupOptions;

@CapacitorPlugin(name = "Vapi")
public class VapiPlugin extends Plugin {

    public static final String TAG = "Vapi";

    private static final String ERROR_API_KEY_MISSING = "apiKey must be provided.";
    private static final String ERROR_UNKNOWN_ERROR = "An unknown error has occurred.";

    private VapiImpl implementation;

    @Override
    public void load() {
        try {
            implementation = new VapiImpl(this);
        } catch (Exception exception) {
            Logger.error(VapiPlugin.TAG, exception.getMessage(), exception);
        }
    }

    @PluginMethod
    public void setup(PluginCall call) {
        try {
            String apiKey = call.getString("apiKey");
            if (apiKey == null) {
                call.reject(ERROR_API_KEY_MISSING);
                return;
            }

            SetupOptions options = new SetupOptions(apiKey);

            implementation.setup(options);
            call.resolve();
        } catch (Exception exception) {
            rejectCall(call, exception);
        }
    }

    private void rejectCall(PluginCall call, Exception exception) {
        String message = exception.getMessage();
        message = (message != null) ? message : ERROR_UNKNOWN_ERROR;
        Logger.error(TAG, message, exception);
        call.reject(message);
    }
}

File: packages/vapi/ios/Sources/VapiPlugin/VapiPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(VapiPlugin)
public class VapiPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "VapiPlugin"
    public let jsName = "Vapi"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "isMuted", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "say", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setMuted", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setup", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "start", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stop", returnType: CAPPluginReturnPromise)
    ]
    private var implementation: VapiImpl?

    override public func load() {
        implementation = VapiImpl(plugin: self)
    }

    @objc func isMuted(_ call: CAPPluginCall) {
        Task {
            do {
                try await implementation?.isMuted()
            } catch {
                rejectCall(call, error)
            }
        }
    }

    @objc func say(_ call: CAPPluginCall) {
        Task {
            do {
                guard let message = call.getString("message") else {
                    rejectCall(call, CustomError.messageMissing)
                    return
                }

                let options = SayOptions(message: message)

                try await implementation?.say(options)
                call.resolve()
            } catch {
                rejectCall(call, error)
            }
        }
    }

    @objc func setMuted(_ call: CAPPluginCall) {
        Task {
            do {
                guard let muted = call.getBool("muted") else {
                    rejectCall(call, CustomError.mutedMissing)
                    return
                }

                let options = SetMutedOptions(muted: muted)

                try await implementation?.setMuted(options)
                call.resolve()
            } catch {
                rejectCall(call, error)
            }
        }
    }

    @objc func setup(_ call: CAPPluginCall) {
        guard let apiKey = call.getString("apiKey") else {
            rejectCall(call, CustomError.apiKeyMissing)
            return
        }

        let options = SetupOptions(apiKey: apiKey)

        implementation?.setup(options)
        call.resolve()
    }

    @objc func start(_ call: CAPPluginCall) {
        Task {
            do {
                guard let assistantId = call.getString("assistantId") else {
                    rejectCall(call, CustomError.assistantIdMissing)
                    return
                }

                let options = StartOptions(assistantId: assistantId)

                try await implementation?.start(options)
                call.resolve()
            } catch {
                rejectCall(call, error)
            }
        }
    }

    @objc func stop(_ call: CAPPluginCall) {
        Task {
            do {
                try await implementation?.stop()
                call.resolve()
            } catch {
                rejectCall(call, error)
            }
        }
    }
    
    func notifyCallEndListeners() {
        notifyListeners("callEnd", data: nil)
    }
    
    func notifyCallStartListeners() {
        notifyListeners("callStart", data: nil)
    }
    
    func notifyConversationUpdateListeners(event: ConversationUpdateEvent) {
        if let event = event.toJSObject() as? JSObject {
            notifyListeners("conversationUpdate", data: event)
        }
    }
    
    func notifyErrorListeners(event: ErrorEvent) {
        if let event = event.toJSObject() as? JSObject {
            notifyListeners("error", data: event)
        }
    }
    
    func notifySpeechUpdateListeners(event: SpeechUpdateEvent) {
        if let event = event.toJSObject() as? JSObject {
            notifyListeners("speechUpdate", data: event)
        }
    }

    private func rejectCall(_ call: CAPPluginCall, _ error: Error) {
        var message = error.localizedDescription
        switch error {
        case let CustomError.apiKeyMissing:
            message = "apiKey must be provided."
        case let CustomError.assistantIdMissing:
            message = "assistantId must be provided."
        case let CustomError.customError(msg):
            message = msg
        case let CustomError.messageMissing:
            message = "message must be provided."
        case let CustomError.mutedMissing:
            message = "muted must be provided."
        case let CustomError.uninitialized:
            message = "Vapi is not initialized. Please call setup() first."
        default:
            break
        }
        call.reject(message, nil, error)
    }
}
